packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,262 @@
+     * @notice withdraw ETH deposited into crab
+     * @dev this will give the sender ZBEPR token as ownership for the ETH deposited in Euler pool
+     * @param _zenBullAmount ZenBull amount to redeem
+     * @param _maxEthForWPowerPerp max ETH to pay for flashswapped oSQTH amount
+     */
+    function emergencyWithdrawEthFromCrab(uint256 _zenBullAmount, uint256 _maxEthForWPowerPerp)
+        external
+    {
+        IERC20(zenBull).transferFrom(msg.sender, address(this), _zenBullAmount);
+
+        uint256 circulatingTotalSupply =
+            IERC20(zenBull).totalSupply().sub(redeemedZenBullAmountForCrabWithdrawal);
+        uint256 crabToRedeem = _zenBullAmount.wdiv(circulatingTotalSupply).wmul(
+            IZenBullStrategy(zenBull).getCrabBalance()
+        );
+        (, uint256 wPowerPerpInCrab) = IZenBullStrategy(zenBull).getCrabVaultDetails();
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+
+        _mint(msg.sender, _zenBullAmount);
+
+        redeemedZenBullAmountForCrabWithdrawal =
+            redeemedZenBullAmountForCrabWithdrawal.add(_zenBullAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            3000,
+            wPowerPerpToRedeem,
+            _maxEthForWPowerPerp,
+            uint8(FLASH_SOURCE.EMERGENCY_WITHDRAW_ETH_FROM_CRAB),
+            abi.encodePacked(crabToRedeem, wPowerPerpToRedeem)
+        );
+
+        uint256 payout = IERC20(weth).balanceOf(address(this));
+        IWETH9(weth).withdraw(payout);
+        payable(msg.sender).sendValue(payout);
+
+        emit Withdraw(
+            msg.sender, _zenBullAmount, crabToRedeem, wPowerPerpToRedeem, payout, _zenBullAmount
+        );
+    }
+
+    /**
+     * @notice repay a portion of Euler debt and withdraw WETH collateral to this contract based on ZenBullEulerRecovery amount
+     * @dev this will revert of WETH to withdraw from Euler is greater than max, or limit price breach the tolerance %
+     * @dev if all ZenBullEulerRecovery are burnt, the ETH withdrawl will be activated
+     * @param _ratio ratio of WETH to withdraw from total WETH in Euler in 18 decimals (e.g 2e17 => 0.2 => 20%)
+     * @param _limitPriceUsdcPerEth ETH limit price
+     * @param _poolFee ETH/USDC Uni v3 pool fee
+     */
+    function emergencyRepayEulerDebt(uint256 _ratio, uint256 _limitPriceUsdcPerEth, uint24 _poolFee)
+        external
+    {
+        uint256 usdcToRepay = _ratio.wmul(IEulerDToken(dToken).balanceOf(zenBull));
+        uint256 wethToWithdraw = _ratio.wmul(IEulerEToken(eToken).balanceOfUnderlying(zenBull));
+
+        require(
+            wethToWithdraw <= MAX_WETH_PER_DEBT_REPAY,
+            "WETH to withdraw is greater than max per repay"
+        );
+
+        uint256 ethUsdcPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, 420, false);
+
+        require(
+            _limitPriceUsdcPerEth <= ethUsdcPrice.wmul((ONE.add(LIMIT_PRICE_TOLERANCE))),
```

Comment:
This isn't right

ethUsdcPrice is something like 1000 usdc per 1 eth
Limit price tolerance is something like 1%

We are selling eth to buy usdc, so we need to make sure the limit price is above

``` _limitPriceUsdcPerEth >= ethUsdcPrice.wmul((ONE.sub(LIMIT_PRICE_TOLERANCE)))```

minLimitPrice shouild be something like 1000 * (100%-1%) = 990
so if I used a limit price of 980 it should revert, but 990 should be allowed